libraries/browser-tracker-core/src/tracker/cookie_storage.ts

```diff
@@ -0,0 +1,174 @@
+import { cookie, deleteCookie } from '../helpers';
+
+
+/**
+ * Cookie storage interface for reading and writing cookies.
+ */
+export interface CookieStorage {
+  /**
+   * Get the value of a cookie
+   *
+   * @param name - The name of the cookie
+   * @returns The cookie value
+   */
+  getCookie(name: string): string;
+
+  /**
+   * Set a cookie
+   *
+   * @param name - The cookie name (required)
+   * @param value - The cookie value
+   * @param ttl - The cookie Time To Live (seconds)
+   * @param path - The cookies path
+   * @param domain - The cookies domain
+   * @param samesite - The cookies samesite attribute
+   * @param secure - Boolean to specify if cookie should be secure
+   * @returns string The cookies value
+   */
+  setCookie(
+    name: string,
+    value?: string,
+    ttl?: number,
+    path?: string,
+    domain?: string,
+    samesite?: string,
+    secure?: boolean
+  ): boolean;
+
+  /**
+   * Delete a cookie
+   *
+   * @param name - The cookie name
+   * @param domainName - The cookie domain name
+   * @param sameSite - The cookie same site attribute
+   * @param secure - Boolean to specify if cookie should be secure
+   */
+  deleteCookie(name: string, domainName?: string, sameSite?: string, secure?: boolean): void;
+
+  /**
+   * Clear the cookie storage cache (does not delete any cookies)
+   */
+  clearCache(): void;
+}
+
+interface Cookie {
+  getValue: () => string;
+  setValue: (value?: string, ttl?: number, path?: string, domain?: string, samesite?: string, secure?: boolean) => boolean;
+  deleteValue: (domainName?: string, sameSite?: string, secure?: boolean) => void;
+}
+
+function newCookie(name: string): Cookie {
+  let cachedValue: string | undefined;
+  let setCookieTimer: ReturnType<typeof setTimeout> | undefined;
+  const debounceTimeout = 10;
+
+  function scheduleClearCache(ttl: number) {
+    setCookieTimer = setTimeout(() => {
+      cachedValue = undefined;
+    }, ttl * 1000);
+  }
+
+  function getValue(): string {
+    // Note: we can't cache the cookie value as we don't know the expiration date
+    return cachedValue ?? cookie(name);
+  }
+
+  function setValue(value?: string, ttl?: number, path?: string, domain?: string, samesite?: string, secure?: boolean): boolean {
+    cachedValue = value;
+
+    // debounce setting the cookie
+    if (setCookieTimer !== undefined) {
+      clearTimeout(setCookieTimer);
+    }
+    setCookieTimer = setTimeout(() => {
+      setCookieTimer = undefined;
+      cookie(name, value, ttl, path, domain, samesite, secure);
+
+      if (ttl !== undefined) {
+        scheduleClearCache(ttl);
+      }
+    }, debounceTimeout);
```

Comment:
I am just thinking... Why aren't we using the requestAnimationFrame and setTimeout trick here?
```
requestAnimationFrame(() => {
    setTimeout(() => {
       cookie()...
    }, 0);
  });
```
Since the time is random it might again hit a blocking point when rendering would be critical and INP metrics might still be an issue. `requestAnimationFrame` is a bit safer on that aspect.

Reply:
I think the race risk is already pretty high with this change, adding indeterminate amounts of time before updating the value seems dangerous to values like `eventIndex`.